localstack/services/s3/v3/storage.py

```diff
@@ -0,0 +1,528 @@
+
+    def copy(
+        self,
+        src_bucket: BucketName,
+        src_object: S3Object,
+        dest_bucket: BucketName,
+        dest_object: S3Object,
+    ) -> EphemeralS3StoredObject:
+        if src_bucket == dest_bucket and src_object.key == dest_object.key:
+            return self.open(src_bucket, src_object)
+
+        src_stored_object = self.open(src_bucket, src_object)
+        dest_stored_object = self.open(dest_bucket, dest_object)
+
+        dest_stored_object.write(src_stored_object)
+
+        return dest_stored_object
+
+    def get_multipart(
+        self, bucket: BucketName, s3_multipart: S3Multipart
+    ) -> EphemeralS3StoredMultipart:
+        upload_key = self._resolve_upload_directory(bucket, s3_multipart.id)
+        if not (multipart := self._get_multipart(bucket, upload_key)):
+
+            upload_dir = self._create_upload_directory(bucket, s3_multipart.id)
+
+            multipart = EphemeralS3StoredMultipart(self, s3_multipart, upload_dir)
+            self._filesystem[bucket]["multiparts"][upload_key] = multipart
+
+        return multipart
+
+    def remove_multipart(self, bucket: BucketName, s3_multipart: S3Multipart):
+        if multiparts := self._filesystem.get(bucket, {}).get("multiparts", {}):
+            upload_key = self._resolve_upload_directory(bucket, s3_multipart.id)
+            if multipart := multiparts.pop(upload_key, None):
+                multipart.close()
+
+        # if the bucket is now empty after removing, we can delete the directory
+        if not multiparts and not self._filesystem.get(bucket, {}).get("keys"):
+            self._delete_bucket_directory(bucket)
+
+    def close(self):
+        for bucket in self._filesystem.values():
+            if keys := bucket.get("keys"):
+                for file in keys.values():
+                    file.close()
+                keys.clear()
+
+            if multiparts := bucket.get("multiparts"):
+                for multipart in multiparts.values():
+                    multipart.close()
+                multiparts.clear()
+
+    @staticmethod
+    def _key_from_s3_object(s3_object: S3Object) -> str:
+        return str(hash(f"{s3_object.key}?{s3_object.version_id or 'null'}"))
+
+    def _get_object_file(self, bucket: BucketName, key: str) -> LockedSpooledTemporaryFile | None:
+        return self._filesystem.get(bucket, {}).get("keys", {}).get(key)
+
+    def _get_multipart(self, bucket: BucketName, upload_key: str) -> S3StoredMultipart | None:
+        return self._filesystem.get(bucket, {}).get("multiparts", {}).get(upload_key)
+
+    @staticmethod
+    def _resolve_upload_directory(bucket_name: BucketName, upload_id: MultipartUploadId) -> str:
+        return f"{bucket_name}/{upload_id}"
+
+    def _create_bucket_directory(self, bucket_name: BucketName) -> str:
+        """
+        Create a temporary directory representing a bucket
+        :param bucket_name
+        """
+        tmp_dir = mkdtemp()
```

Comment:
let's try and restructure the class so we can inject a `root_directory` path as dependency into the constructor, that is then used for all filesystem operations. this will help massively with persistence.

Reply:
You mean in the base class? Because this one is not really supposed to be used with persistence? but for testing, sure, it makes a lot of sense